src/main/java/com/cloudbees/plugins/credentials/CredentialsDescriptor.java

```diff
@@ -192,4 +257,343 @@ public String getIconClassName() {
+                                    // skip the name
+                                    index++;
+                                } else {
+                                    // we have gone as far as we can parse the item path structure
+                                    while (index < pathSegments.size()) {
+                                        // remove the remainder
+                                        pathSegments.remove(index);
+                                    }
+                                }
+                            }
+                            Jenkins jenkins = Jenkins.getActiveInstance();
+                            while (!pathSegments.isEmpty()) {
+                                String fullName = StringUtils.join(pathSegments, "/");
+                                Item item = jenkins.getItemByFullName(fullName);
+                                if (item != null) {
+                                    if (type.isInstance(item) && CredentialsProvider.hasStores(item)) {
+                                        // we have a winner
+                                        return type.cast(item);
+                                    }
+                                }
+                                // walk back up and try one level less deep
+                                pathSegments.remove(pathSegments.size() - 1);
+                            }
+                        }
+                    }
+                    // ok we give up, we are not thirsty for more, we'll let "normal" ancestor in path logic continue
+                }
+            }
+            if (type.isInstance(o) && o instanceof ModelObject && CredentialsProvider.hasStores((ModelObject) o)) {
+                return type.cast(o);
+            }
+        }
+        return null;
+
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public FormValidation.CheckMethod getCheckMethod(String fieldName) {
+        // this is an ugly hack to make the @ContextInPath annotation more failsafe
+        // requires that you explicitly call out the checkUrl: checkUrl="${descriptor.getCheckUrl('fieldName')}"
+        FormValidation.CheckMethod method = enhancedCheckMethods.get(fieldName);
+        if (method == null) {
+            method = new EnhancedCheckMethod(this, fieldName);
+            enhancedCheckMethods.put(fieldName, method);
+        }
+        return method;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void calcFillSettings(String field, Map<String, Object> attributes) {
+        if (attributes.containsKey("fillUrl")) {
+            // the user already provided a custom one, get out of the way
+            super.calcFillSettings(field, attributes);
+            return;
+        }
+        // this is an ugly hack to make the @ContextInPath annotation more failsafe
+        super.calcFillSettings(field, attributes);
+        if (attributes.containsKey("fillUrl")) {
```

Comment:
nope... the `calcFillSettings` method will add the attribute.

The idea is to do a no-op if the user supplied the fillUrl, but if the fillUrl was calculated then we add the `$provider` and `$token`